### ISSUE TITLE
Error reported by sanitizers

### ISSUE THREAD

**Issue by areynoua** [ticktac-project/tchecker#54]
Since I want to extend TChecker, I tried to compile with `-fsanitize=undefined,address,leak`. The compilation works, but at runtime I get some errors like "member access within misaligned address", "constructor call on misaligned address". Is there really a problem here? Did I miss something? Do you know which sanitizers you can enable, or other useful warnings you can use? Many thanks.

**Comment by fredher** [comment]
Hi, could you please give more details on your architecture, OS, and the exact modifications you made on the build system? I could not reproduce the error.
Best regards,
Frédéric

**Comment by areynoua** [comment]
Thank you for your quick response.

I am working on a `Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz` with `Linux Mint 21.1`, `g++ 11.3` and `glibc 2.35` (`libc6 2.35`). See below for details on the architecture and package versions.

Here is how I compiled TChecker:

```bash
# This is the only real difference from the INSTALL.md

CXXFLAGS=' -fsanitize=undefined,address,leak'
export CXXFLAGS

# CXXFLAGS is used by make in its default rules to compile C++ code.
# I use GNU Make 4.3, see https://www.gnu.org/software/make/manual/make.html#Catalogue-of-Rules
# “CXXFLAGS: Extra flags to give to the C++ compiler.”
# NOTE that everything works well with CXXFLAGS=' -fsanitize=leak'

sudo apt install g++ cmake flex bison libboost-dev doxygen

# ...
# pull and compile Catch2
# ...
# pull TChecker

mkdir install
mkdir build
cd build
cmake ../repo -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=../install
make -j
make doc
make install

# Then using tck-reach produce the mentionned errors
```

Package versions

```
$ g++ --version
g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

$ cmake --version
cmake version 3.22.1

$ flex --version
flex 2.6.4

$ bison --version
bison (GNU Bison) 3.8.2

$ dpkg -s 'libboost-dev' | grep ^Version
Version: 1.74.0.3ubuntu7

$ doxygen --version
1.9.1

$ dpkg -s 'libc-dev-bin' | grep ^Version
Version: 2.35-0ubuntu3.1
```

Catch2 was compiled from their GitHub repo with the same CXXFLAGS exported variable.

CPU:

```
$ lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  4
  On-line CPU(s) list:   0-3
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz
    CPU family:          6
    Model:               142
    Thread(s) per core:  2
    Core(s) per socket:  2
    Socket(s):           1
    Stepping:            9
    CPU max MHz:         3500,0000
    CPU min MHz:         400,0000
    BogoMIPS:            5399.81
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art
                          arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid ss
                         e4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_sh
                         adow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dt
                         herm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   64 KiB (2 instances)
  L1i:                   64 KiB (2 instances)
  L2:                    512 KiB (2 instances)
  L3:                    3 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-3
Vulnerabilities:         
  Itlb multihit:         KVM: Mitigation: VMX disabled
  L1tf:                  Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
  Mds:                   Mitigation; Clear CPU buffers; SMT vulnerable
  Meltdown:              Mitigation; PTI
  Mmio stale data:       Mitigation; Clear CPU buffers; SMT vulnerable
  Retbleed:              Mitigation; IBRS
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS Not affected
  Srbds:                 Mitigation; Microcode
  Tsx async abort:       Mitigation; TSX disabled
```

**Comment by fredher** [comment]
Dear Alexis,

Thanks for the details. The main issue comes from the size of type `REFCOUNT` of template `make_shared_t` in file `include/tchecker/utils/shared_objects.hh` line 48. This type should be of the same size as a word on the architecture. 

I will cleanly fix that later, but as a workaround, you can define:
```
template <class T, class REFCOUNT = std::size_t, std::size_t RESERVED = 1> class make_shared_t final : public T {
```
notice the new definition of type REFCOUNT: `std::size_t` instead of `unsigned int`.

I have done a few tests and this seems to fix all errors but one. There is a negative left shift issue that I have not investigated yet. It occurs after termination of `tck-reach`, so it won't bother you too much hopefully. I will fix it as soon as possible. 

Best regards,
Frédéric

**Comment by fredher** [comment]
Hi,
I have fixed the issue with unaligned addresses mentionned in my previous comment. There is still an issue with tchecker::make_array_t. I am working on it and let you know once it is done.
I have also fixed another issue pointed by the sanitizers: the DBM library was relying on compiler-dependent features. The new implementation is slightly slower but safe. The old implementation can still be used by setting TCHECKER_DBM_UNSAFE with cmake.
Best regards,
Frédéric

**Comment by areynoua** [comment]
Here are the errors I get on the same architecture, with [the last version (Jan 31)](https://github.com/ticktac-project/tchecker/tree/d8f3be12ab24267292e557dc20b53db1f8b62004), when running the tests:

```
include/tchecker/utils/pool.hh:387:14:           runtime error: store to             misaligned address 0x7.......... for type 'refcount_t',           which requires 8 byte alignment
include/tchecker/utils/shared_objects.hh:167:15: runtime error: load of              misaligned address 0x7.......... for type 'refcount_t',           which requires 8 byte alignment
include/tchecker/utils/shared_objects.hh:202:15: runtime error: store to             misaligned address 0x7.......... for type 'refcount_t',           which requires 8 byte alignment
include/tchecker/utils/shared_objects.hh:84:5:   runtime error: constructor call on  misaligned address 0x7.......... for type 'struct make_shared_t', which requires 8 byte alignment
include/tchecker/utils/shared_objects.hh:199:28: runtime error: member access within misaligned address 0x7.......... for type 'struct make_shared_t', which requires 8 byte alignment
```

Here are the flags used to compile:
```
-fsanitize=undefined,address,leak -std=c++17 -O0 -Wall -g -g3  -std=gnu++17
```

If I understand correctly, the test on [shared_object:51](https://github.com/ticktac-project/tchecker/blob/d8f3be12ab24267292e557dc20b53db1f8b62004/include/tchecker/utils/shared_objects.hh#L51) is somehow not sufficient. It looks like the condition does not ensure good alignment in the rest of the code.

My first guess would be that the `new char[]` on [line 97](https://github.com/ticktac-project/tchecker/blob/d8f3be12ab24267292e557dc20b53db1f8b62004/include/tchecker/utils/shared_objects.hh#L97) should be `align`ed`as(REFCOUNT)`. This can be done by changing the line into `char * ptr = new(std::align_val_t{alignof(REFCOUNT)}) char[alloc_size];`. However, I do not know how to call the good delete operator, since we do not have “placement delete”s in C++17: In `destruct_and_deallocate`, calling `::operator delete(p, std::align_val_t{alignof(REFCOUNT)});` is not sufficient.
